### PR TITLE
Incorporating feedback on 119341

### PR DESCRIPTION
**Pull request by divyasri537**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Incorporating feedback on PR #119341

#### Which issue(s) this PR fixes:

Fixes #119341

#### Does this PR introduce a user-facing change?

```release-note
Incorporating feedback on PR #119341
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (ivelichkovich) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:172]: I don't think we want to break here either, we probably want to flow through to the end

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (ivelichkovich) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:171]: did you look into checking `err` instead of ctx.Err?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go
Comment (ivelichkovich) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:177]: same thing with break here

**Review by ivelichkovich — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (divyasri537) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:171]: Yes I checked and added the comments in the old PR https://github.com/kubernetes/kubernetes/pull/119341

Comment (divyasri537) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:171]: The err that is from callAttrMutatingHook method [[ref](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go#L165)] based on the provided ctx . The err.Reason returns customized error response that picks up the ctx error so I don’t think anything wrong by checking this way ctx.Err() == context.Canceled as it is similar to checking the err obj.

Also The err.Reason returns
%!(EXTRA string=failed to call webhook: Post "https://127.0.0.1:21375/validate?timeout=2s": context canceled) which is a customized error and may break if someone modifies the error message so I think its better to go with checking ctx.Err().

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (divyasri537) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:172]: I have added break statement here, so the metric counter in next steps wont pick the canceled context.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (ivelichkovich) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:172]: Why did we get rid of the log?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (ivelichkovich) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:204]: ah I see its moved down here

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (ivelichkovich) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:172]: maybe add a comment around what this is for

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (ivelichkovich) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:204]: nit: "Context canceled"

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go
Comment (ivelichkovich) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:177]: same here add comment

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go
Comment (ivelichkovich) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:204]: so here just doing 

```
errCh <- apierrors.NewInternalError(err)
return
```
is not necessarily the same behavior. The `if ignoreClientCallFailures` block uses utilruntime.HandlerError while the other block does the `errCh <- apierrors.NewInternalError(err)`

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (ivelichkovich) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:206]: same here, not necessarily the same behavior. The blocks for `ignoreClientCallFailures` and outside that block handle the error differently

**Review by ivelichkovich — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (divyasri537) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:206]: Modified in the warning message.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (divyasri537) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:208]: utilruntime.HandleError(callErr) will returns "failed calling webhook "eks-aws-auth-configmap-validation-webhook.amazonaws.com": failed to call webhook: Post "https://127.0.0.1:21375/mutate?timeout=2s": context canceled" so keeping this one.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go
Comment (ivelichkovich) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:178]: update comment, it doesn't only apply to failopen metrics

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (ivelichkovich) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:172]: this doesn't only apply to failopen metric

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (ivelichkovich) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:208]: So @liggitt I think if context is cancelled it makes sense to just return this error. Previous behavior was it would `continue` `if ignoreClientCallFailures {` but we can probably just return and bail out

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go
Comment (ivelichkovich) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:205]: I think it probably makes sense to always do `errCh <- apierrors.NewInternalError(err)` for context cancelled but will differ to @liggitt

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:178]: skipping the webhook rejection metric for cancelled requests seems ok, but do we really want to lose visibility to the fact that the webhook was called *at all* and the latency we hit for canceled requests?

I was hoping we could revert back to the state at 5e5b3029f3bbfc93c3569f07ad300a5c6057fc58 and narrowly skip the `ObserveWebhookRejection` and `ObserveWebhookFailOpen`/`addFailedOpenAnnotation` bits when the reason was a context cancel

something like this:

```diff
$ git diff -w 5e5b3029f3bbfc93c3569f07ad300a5c6057fc58 HEAD -- staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
diff --git a/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go b/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
index c1d1ca6ff6b..61b17932df1 100644
--- a/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
+++ b/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
@@ -20,6 +20,7 @@ package mutating
 
 import (
        "context"
+       "errors"
        "fmt"
        "time"
 
@@ -164,14 +165,18 @@ func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr admission.Attrib
                annotator := newWebhookAnnotator(versionedAttr, round, i, hook.Name, invocation.Webhook.GetConfigurationName())
                changed, err := a.callAttrMutatingHook(ctx, hook, invocation, versionedAttr, annotator, o, round, i)
                ignoreClientCallFailures := hook.FailurePolicy != nil && *hook.FailurePolicy == admissionregistrationv1.Ignore
+
                rejected := false
                if err != nil {
                        switch err := err.(type) {
                        case *webhookutil.ErrCallingWebhook:
                                if !ignoreClientCallFailures {
                                        rejected = true
+                                       // Ignore context cancelled from webhook metrics
+                                       if !errors.Is(err.Reason, context.Canceled) {
                                                admissionmetrics.Metrics.ObserveWebhookRejection(ctx, hook.Name, "admit", string(versionedAttr.Attributes.GetOperation()), admissionmetrics.WebhookRejectionCallingWebhookError, int(err.Status.ErrStatus.Code))
                                        }
+                               }
                                admissionmetrics.Metrics.ObserveWebhook(ctx, hook.Name, time.Since(t), rejected, versionedAttr.Attributes, "admit", int(err.Status.ErrStatus.Code))
                        case *webhookutil.ErrWebhookRejection:
                                rejected = true
@@ -199,9 +204,14 @@ func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr admission.Attrib
 
                if callErr, ok := err.(*webhookutil.ErrCallingWebhook); ok {
                        if ignoreClientCallFailures {
+                               // Ignore context cancelled from webhook metrics
+                               if errors.Is(callErr.Reason, context.Canceled) {
+                                       klog.Warningf("Context canceled when calling webhook %v", hook.Name)
+                               } else {
                                        klog.Warningf("Failed calling webhook, failing open %v: %v", hook.Name, callErr)
                                        admissionmetrics.Metrics.ObserveWebhookFailOpen(ctx, hook.Name, "admit")
                                        annotator.addFailedOpenAnnotation()
+                               }
 
                                utilruntime.HandleError(callErr)
```

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:178]: (same comments apply to the validating dispatcher)

Comment (ivelichkovich) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:178]: Sure that makes sense

**Comment by liggitt** [comment]
/hold

please squash down to a single commit before merge

**Comment by divyasri537** [comment]
> /hold
> 
> please squash down to a single commit before merge

@liggitt Squashed to one commit.

**Pull request merged**
